packages/yarnpkg-fslib/sources/ZipFS.ts

```diff
@@ -1478,4 +1389,67 @@ export class ZipFS extends BasePortableFakeFS {
     const interval = setInterval(() => {}, 24 * 60 * 60 * 1000);
     return {on: () => {}, close: () => {clearInterval(interval);}};
   }
+
+  watchFile(p: PortablePath, cb: WatchFileCallback): StatWatcher;
+  watchFile(p: PortablePath, opts: WatchFileOptions, cb: WatchFileCallback): StatWatcher;
+  watchFile(p: PortablePath, a: WatchFileOptions | WatchFileCallback, b?: WatchFileCallback) {
+    const resolvedP = this.resolveFilename(`open '${p}'`, p);
+
+    let bigint: boolean;
+    let persistent: boolean;
+    let interval: number;
+
+    let listener: WatchFileCallback;
+
+    switch (typeof a) {
+      case `function`: {
+        bigint = false;
+        persistent = true;
+        interval = 5007;
+
+        listener = a;
+      } break;
+
+      default: {
+        ({
+          bigint = false,
+          persistent = true,
+          interval = 5007,
+        } = a);
+
+        listener = b!;
+      } break;
+    }
```

Comment:
Can we abstract all this logic, since it would likely be the same for all FS watchers? For instance, can we do something like this:

```ts
const resolvedP = this.resolveFilename(`open '${p}'`, p);
const {watcher, unwatch} = createStatWatcher(this, resolvedP, a, b);
```

Then let the `create` method decide how to interpret `a` and `b`

Reply:
I was initially thinking about doing that, but the problem is that both methods require access to the internal private state of the `ZipFS` (the `statWatchers` map). An option would be to pass a reference to the `createStatWatcher` function. Another option would be to do something like:
```ts
export type StatWatchable<P extends Path> = {
  statWatchers: Map<P, CustomStatWatcher<P>>;
}

// In CustomStatWatcher we only accept a StatWatchable FakeFS instead of a regular FakeFS

// Then in ZipFS
export class ZipFS extends FakeFS<PortablePath> implements StatWatchable<PortablePath> {...}
```

Another question: Where would `unwatch` be stored? Should we have another Map mapping resolved paths to unwatchers?

Reply:
> I was initially thinking about doing that, but the problem is that both methods require access to the internal private state of the `ZipFS` (the `statWatchers` map)

What about using a WeakMap, so we can keep the `statWatchers` outside of the `ZipFS` implementation? It seems to be a case where the `statWatchers` would *only* be 'public' to the watcher code, so I think it'd be fine to keep the stores there.

> Another question: Where would `unwatch` be stored? Should we have another Map mapping resolved paths to unwatchers?

Depends - either we keep them internal, or we go full "weakmap-based" and we do something like this:

```ts
class ZipFS {
  saveAndClose() {
    disposeAllStatWatcher(this);
  }
  unwatchFile() {
    disposeStatWatcher(this, p);
  }
}
```

I think this approach is worth experimenting, if only because it makes this part of the code 100% independent from ZipFS, so it can be reused for other archive formats if it ever comes to that.